pennylane_cirq/simulator_device.py

```diff
@@ -177,8 +182,75 @@ def generate_samples(self):
         # Bring measurements to a more managable form, but keep True/False as values for now
         # They will be changed in the measurement routines where the observable is available
         return np.array(
-            [
-                self._result.measurements[str(wire)].flatten()
-                for wire in range(self.num_wires)
-            ]
+            [self._result.measurements[str(wire)].flatten() for wire in range(self.num_wires)]
         ).T.astype(int)
+
+
+class MixedStateSimulatorDevice(SimulatorDevice):
+    r"""Cirq mixed-state simulator device for PennyLane.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. Shots need
+            to >= 1. In analytic mode, shots indicates the number of entries
+            that are returned by device.sample.
+        analytic (bool): Indicates that expectation values and variances should
+            be calculated analytically. Defaults to `True`.
+        qubits (List[cirq.Qubit]): a list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of `cirq.LineQubit` instances is created.
+    """
+    name = "Cirq Mixed-State Simulator device for PennyLane"
+    short_name = "cirq.mixedsimulator"
+
+    _mixed_sim_operation_map = {
+        "BitFlip": CirqOperation(cirq.bit_flip),
+        "PhaseFlip": CirqOperation(cirq.phase_flip),
+        "PhaseDamp": CirqOperation(cirq.phase_damp),
+        "AmplitudeDamp": CirqOperation(cirq.amplitude_damp),
+        "Depolarize": CirqOperation(cirq.depolarize),
+    }
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None):
+        self._operation_map = dict(self._operation_map, **self._mixed_sim_operation_map)
+        super().__init__(wires, shots, analytic, qubits)
+
+        self._simulator = cirq.DensityMatrixSimulator()
+
+        self._initial_state = None
+        self._result = None
+        self._state = None
+
+    def _apply_basis_state(self, basis_state_operation):
+        super()._apply_basis_state(basis_state_operation)
+        self._initial_state = self._convert_to_density_matrix(self._initial_state)
+
+    def _apply_qubit_state_vector(self, qubit_state_vector_operation):
+        super()._apply_qubit_state_vector(qubit_state_vector_operation)
+        self._initial_state = self._convert_to_density_matrix(self._initial_state)
+
+    def _convert_to_density_matrix(self, state_vec):
+        """Convert ``state_vec`` into a density matrix."""
+        dim = 2 ** self.num_wires
+        return np.kron(state_vec, state_vec.conj()).reshape((dim, dim))
+
+    def _get_state_from_cirq(self, result):
+        """Extract the state array from a Cirq TrialResult"""
+        return np.array(result.final_density_matrix)
```

Comment:
Same comment here re: self usage. Unless it's likely that other devices might need `self` in the future?